src/matrix/layout_info.cpp

```diff
@@ -0,0 +1,84 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include "dlaf/matrix/layout_info.h"
+#include "dlaf/util_math.h"
+
+namespace dlaf {
+namespace matrix {
+LayoutInfo::LayoutInfo(GlobalElementSize size, TileElementSize block_size, SizeType tile_ld,
+                       std::size_t tile_offset_row, std::size_t tile_offset_col)
+    : size_(size), nr_tiles_(0, 0), block_size_(block_size), ld_tile_(tile_ld),
+      tile_offset_row_(tile_offset_row), tile_offset_col_(tile_offset_col) {
+  using util::size_t::sum;
+  using util::size_t::mul;
+
+  if (size_.rows() < 0 || size_.cols() < 0)
+    throw std::invalid_argument("Error: Invalid Matrix size");
+  if (block_size_.rows() < 1 || block_size_.cols() < 1)
+    throw std::invalid_argument("Error: Invalid Block size");
+
+  nr_tiles_ = {util::ceilDiv(size_.rows(), block_size_.rows()),
+               util::ceilDiv(size_.cols(), block_size_.cols())};
+
+  if (size_.rows() == 0 || size_.cols() == 0) {
+    if (ld_tile_ < 1) {
+      throw std::invalid_argument("Error: Invalid Leading Dimension");
+    }
+    if (tile_offset_row_ < 1) {
+      throw std::invalid_argument("Error: Invalid Tile Row Offset");
+    }
+    if (tile_offset_col_ < 1) {
+      throw std::invalid_argument("Error: Invalid Tile Col Offset");
+    }
+  }
+  else {
+    SizeType last_rows = size_.rows() - block_size_.rows() * (nr_tiles_.rows() - 1);
+
+    if (ld_tile_ < block_size_.rows()) {
+      throw std::invalid_argument("Error: Invalid Leading Dimension");
+    }
+    if (tile_offset_row_ < static_cast<std::size_t>(block_size_.rows())) {
+      throw std::invalid_argument("Error: Invalid Tile Row Offset");
+    }
+    if (tile_offset_row_ < minTileSize(block_size_) &&
+        static_cast<std::size_t>(ld_tile_) <
+            sum(mul((nr_tiles_.rows() - 1), tile_offset_row_), last_rows)) {
+      throw std::invalid_argument("Error: Invalid Leading Dimension & Tile Row Offset combination");
+    }
+    if (tile_offset_col_ <
+        mul(nr_tiles_.rows() - 1, tile_offset_row_) + minTileSize({last_rows, block_size_.cols()})) {
```

Comment:
I would try to improve readability by extracting a few variables with a self-explaining name.